src/build.rs

```diff
@@ -4,4 +4,9 @@ fn main() {
     if var("CARGO_FEATURE_DEFLATE_MINIZ").is_ok() {
         println!("cargo:warning=Feature `deflate-miniz` is deprecated; replace it with `deflate`");
     }
+    #[cfg(not(any(feature = "sync", feature = "tokio")))]
+    compile_error!("Missing Required feature");
+
+    #[cfg(all(feature = "sync", feature = "tokio"))]
+    compile_error!("The features sync and tokio cannot be used together")
```

Comment:
I'm thinking of pushing the old synchronous code in a module of read named sync and then the async code would be in a tokio module, this would avoid conflicts between the codebases, by the end of the week i'm probably gonna push a PR for that since it doesn't require any feature or ci changes.

Reply:
That sounds good, but will there be a shared-core module as well?

Reply:
The idea is that the structs and enums can be shared but the actual logic can be separate. This should avoid import conflicts and if there are code changes in the sync part it should not conflict with the async part. Do u think this or the macro one is better?

Reply:
I favor sharing as much code as possible, based on the Don't Repeat Yourself principle. But the modular design should be compatible with that: define the macros in the core module, and invoke them in the sync and tokio macros.

Reply:
@SapryWenInera Could you please prioritize this issue for a fix, so that I can run CI on your work in progress and use the results to estimate how much longer this PR will take?